File: baba-chatbot-net/src/Baba.Chatbot.Domain/Entities/Vehicle.cs
namespace Baba.Chatbot.Domain.Entities;

/// <summary>
/// Entidad de dominio que representa un vehículo disponible en el catálogo
/// Incluye características técnicas, especificaciones, estado y certificación
/// </summary>
public class Vehicle
{
    public string Id { get; private set; }
    public string Brand { get; private set; }
    public string Model { get; private set; }
    public int Year { get; private set; }
    public string Version { get; private set; }
    public decimal Price { get; private set; }
    public int Mileage { get; private set; }
    public string Transmission { get; private set; }
    public string FuelType { get; private set; }
    public string Color { get; private set; }
    public int Doors { get; private set; }
    public int Seats { get; private set; }
    public string Engine { get; private set; }
    public int Horsepower { get; private set; }
    public List<string> Features { get; private set; } = new();
    public string Location { get; private set; }
    public VehicleStatus Status { get; private set; }
    public decimal CertificationScore { get; private set; }
    public int PreviousOwners { get; private set; }
    public bool HasAccidents { get; private set; }
    
    // Dimensiones
    public int? Length { get; private set; }
    public int? Width { get; private set; }
    public int? Height { get; private set; }
    
    // Características adicionales
    public bool? HasBluetooth { get; private set; }
    public bool? HasCarPlay { get; private set; }
    
    public string StockId { get; private set; } = string.Empty;

    /// <summary>
    /// Constructor privado para encapsular creación
    /// </summary>
    private Vehicle()
    {
        Id = string.Empty;
        Brand = string.Empty;
        Model = string.Empty;
        Version = string.Empty;
        Transmission = string.Empty;
        FuelType = string.Empty;
        Color = string.Empty;
        Engine = string.Empty;
        Location = string.Empty;
    }

    /// <summary>
    /// Crea una nueva instancia de vehículo con los datos básicos requeridos
    /// </summary>
    public static Vehicle Create(
        string id,
        string brand,
        string model,
        int year,
        decimal price)
    {
        return new Vehicle
        {
            Id = id,
            Brand = brand,
            Model = model,
            Year = year,
            Price = price,
            Status = VehicleStatus.Available
        };
    }
    
    /// <summary>
    /// Actualiza detalles opcionales del vehículo como versión, kilometraje y características
    /// </summary>
    public void UpdateDetails(
        string? version = null,
        int? mileage = null,
        int? length = null,
        int? width = null,
        int? height = null,
        bool? hasBluetooth = null,
        bool? hasCarPlay = null,
        string? stockId = null)
    {
        if (version != null) Version = version;
        if (mileage.HasValue) Mileage = mileage.Value;
        if (length.HasValue) Length = length;
        if (width.HasValue) Width = width;
        if (height.HasValue) Height = height;
        if (hasBluetooth.HasValue) HasBluetooth = hasBluetooth;
        if (hasCarPlay.HasValue) HasCarPlay = hasCarPlay;
        if (stockId != null) StockId = stockId;
    }

    /// <summary>
    /// Verifica si el vehículo está disponible para venta
    /// </summary>
    public bool IsAvailable() => Status == VehicleStatus.Available;

    /// <summary>
    /// Marca el vehículo como reservado si está disponible
    /// </summary>
    public void Reserve()
    {
        if (Status == VehicleStatus.Available)
        {
            Status = VehicleStatus.Reserved;
        }
    }

    public void Sell()
    {
        Status = VehicleStatus.Sold;
    }
}

/// <summary>
/// Estado actual de disponibilidad de un vehículo
/// </summary>
public enum VehicleStatus
{
    Available,
    Reserved,
    Sold,
    Maintenance
}

File: baba-chatbot-python/src/baba_chatbot/domain/entities/vehicle.py
"""
Entidad de dominio que representa un vehículo disponible en el catálogo.
Incluye características técnicas, especificaciones, estado y certificación.
"""
from enum import Enum
from typing import List, Optional
from dataclasses import dataclass, field


class VehicleStatus(Enum):
    """Estado actual de disponibilidad de un vehículo"""
    AVAILABLE = "available"
    RESERVED = "reserved"
    SOLD = "sold"
    MAINTENANCE = "maintenance"


@dataclass
class Vehicle:
    """
    Entidad de dominio que representa un vehículo disponible en el catálogo.
    """
    id: str
    brand: str
    model: str
    year: int
    price: float
    version: str = ""
    mileage: int = 0
    transmission: str = ""
    fuel_type: str = ""
    color: str = ""
    doors: int = 0
    seats: int = 0
    engine: str = ""
    horsepower: int = 0
    features: List[str] = field(default_factory=list)
    location: str = ""
    status: VehicleStatus = VehicleStatus.AVAILABLE
    certification_score: float = 0.0
    previous_owners: int = 0
    has_accidents: bool = False
    
    # Dimensiones
    length: Optional[int] = None
    width: Optional[int] = None
    height: Optional[int] = None
    
    # Características adicionales
    has_bluetooth: Optional[bool] = None
    has_carplay: Optional[bool] = None
    
    stock_id: str = ""

    @staticmethod
    def create(
        vehicle_id: str,
        brand: str,
        model: str,
        year: int,
        price: float
    ) -> 'Vehicle':
        """
        Crea una nueva instancia de vehículo con los datos básicos requeridos.
        
        Args:
            vehicle_id: Identificador único del vehículo
            brand: Marca del vehículo
            model: Modelo del vehículo
            year: Año del vehículo
            price: Precio del vehículo
            
        Returns:
            Nueva instancia de Vehicle
        """
        return Vehicle(
            id=vehicle_id,
            brand=brand,
            model=model,
            year=year,
            price=price,
            status=VehicleStatus.AVAILABLE
        )

    def update_details(
        self,
        version: Optional[str] = None,
        mileage: Optional[int] = None,
        length: Optional[int] = None,
        width: Optional[int] = None,
        height: Optional[int] = None,
        has_bluetooth: Optional[bool] = None,
        has_carplay: Optional[bool] = None,
        stock_id: Optional[str] = None
    ) -> None:
        """
        Actualiza detalles opcionales del vehículo como versión, kilometraje y características.
        
        Args:
            version: Versión del vehículo
            mileage: Kilometraje del vehículo
            length: Longitud del vehículo en mm
            width: Ancho del vehículo en mm
            height: Altura del vehículo en mm
            has_bluetooth: Si tiene Bluetooth
            has_carplay: Si tiene CarPlay
            stock_id: ID de stock
        """
        if version is not None:
            self.version = version
        if mileage is not None:
            self.mileage = mileage
        if length is not None:
            self.length = length
        if width is not None:
            self.width = width
        if height is not None:
            self.height = height
        if has_bluetooth is not None:
            self.has_bluetooth = has_bluetooth
        if has_carplay is not None:
            self.has_carplay = has_carplay
        if stock_id is not None:
            self.stock_id = stock_id

    def is_available(self) -> bool:
        """Verifica si el vehículo está disponible para venta."""
        return self.status == VehicleStatus.AVAILABLE

    def reserve(self) -> None:
        """Marca el vehículo como reservado si está disponible."""
        if self.status == VehicleStatus.AVAILABLE:
            self.status = VehicleStatus.RESERVED

    def sell(self) -> None:
        """Marca el vehículo como vendido."""
        self.status = VehicleStatus.SOLD

